software/firmware/experimental/custom_font.py

```diff
@@ -0,0 +1,171 @@
+        if font.hmap():
+            self.map = framebuf.MONO_HMSB if font.reverse() else framebuf.MONO_HLSB
+        else:
+            raise ValueError("Font must be horizontally mapped.")
+        self.screenwidth = device.width  # In pixels
+        self.screenheight = device.height
+
+    def print(self, string, x, y, c=1):
+        """Print the string using the x, y coordinates as the upper-left corner of the text.
+        With c=0, the text is display as black text on white background.
+        """
+        for char in string:
+            if char == "\n":  # line breaks are ignored
+                return
+            glyph, char_height, char_width = self.font.get_ch(char)
+            buf = bytearray(glyph)
+            if c != 1:
+                for i, v in enumerate(buf):
+                    buf[i] = 0xFF & ~ v
+            fbc = framebuf.FrameBuffer(buf, char_width, char_height, self.map)
+            self.device.blit(fbc, x, y)
+            x += char_width
+
+    def string_len(self, string):
+        """Returns the length of string in pixels."""
+        n = 0
+        for char in string:
+            n += self._char_len(char)
+        return n
+
+    def _char_len(self, char):
+        """Returns the length of char in pixels."""
+        if char == "\n":
+            char_width = 0
+        else:
+            _, _, char_width = self.font.get_ch(char)
+        return char_width
+
+
+class Display(SSD1306_I2C):
+    """A class for drawing graphics and text to the OLED.
+
+    The OLED Display works by collecting all the applied commands and only
+    updates the physical display when ``oled.show()`` is called. This allows
+    you to perform more complicated graphics without slowing your program, or
+    to perform the calculations for other functions, but only update the
+    display every few steps to prevent lag.
+
+    To clear the display, simply fill the display with the colour black by using ``oled.fill(0)``
+
+    More explanations and tips about the display can be found in the oled_tips file
+    `oled_tips.md <https://github.com/Allen-Synthesis/EuroPi/blob/main/software/oled_tips.md>`_
+    """
+
+    def __init__(
+            self,
+            sda,
+            scl,
+            width=OLED_WIDTH,
+            height=OLED_HEIGHT,
+            channel=I2C_CHANNEL,
+            freq=I2C_FREQUENCY,
+            default_font=None  # by default will use the monospaced 8x8 font
+    ):
+        i2c = I2C(channel, sda=Pin(sda), scl=Pin(scl), freq=freq)
+        self.width = width
+        self.height = height
+        self.writers = {}  # re-usable large font writer instances
+        self.default_font = default_font
+        if len(i2c.scan()) == 0:
+            if not TEST_ENV:
+                raise Exception(
+                    "EuroPi Hardware Error:\nMake sure the OLED display is connected correctly"
+                )
+        super().__init__(self.width, self.height, i2c)
+
+    def writer(self, font):
```

Comment:
[discussion] I think that this method should be 'private', as in `_writer`. I don't think that it is intended to be used by clients of the `Display` class.

Reply:
Agree